### PR TITLE
Implement --no-redirect for disabling the redirect to a Composer-inst…

### PR DESCRIPTION
**Pull request by DanielRuf**
…alled version

This implements my proposed feature in #5495.

Also `--phpunit-xml` was missing in the command docs.

### CODE REVIEW

File: autoload.php [hunk 1]
```diff
@@ -1,7 +1,6 @@
 <?php
-
 $autoloadFile = './vendor/codeception/codeception/autoload.php';
-if (file_exists('./vendor/autoload.php') && file_exists($autoloadFile) && __FILE__ != realpath($autoloadFile)) {
+if ( ( isset($argv) && isset($argv['--no-redirect']) ) && file_exists('./vendor/autoload.php') && file_exists($autoloadFile) && __FILE__ != realpath($autoloadFile)) {
```
Comment (Naktibalda) [inline autoload.php]: I think that you got this condition backwards, it should be `!isset($argv['--no-redirect'])`.

Comment (DanielRuf) [reply autoload.php]: Oops, sorry. You are right

**Comment by DavertMik** [comment]
I agree this is quite a useful feature, esp for Codeception developers.
Can we call it something better than `no-redirect`?

Like `prefer-composer` or `prefer-local`?

**Comment by DanielRuf** [comment]
Hi @DavertMik,

Thanks for your feedback, I will rename it to `prefer-local`. This should prevent breaking changes. The other way / suggested name would have to be negated.

**Comment by DanielRuf** [comment]
> I will rename it to `prefer-local`

Done.

**Comment by DavertMik** [comment]
@Naktibalda I'm going to merge this, are you ok?

File: src/Codeception/Command/Run.php [hunk 2]
```diff
@@ -61,6 +61,7 @@
  *  --html                Generate html with results (default: "report.html")
  *  --xml                 Generate JUnit XML Log (default: "report.xml")
  *  --phpunit-xml         Generate PhpUnit XML Log (default: "phpunit-report.xml")
+ *  --prefer-local        Prefer a local to a Composer-installed version
```
Comment (Naktibalda) [inline src/Codeception/Command/Run.php]: This description could be better, what is that `local`?

Comment (diverent2) [reply src/Codeception/Command/Run.php]:
```suggestion
 *  --prefer-local        Insist on using a local version over the one defined in composer.json
```

Comment (Naktibalda) [reply src/Codeception/Command/Run.php]: I think that I disagree with usage of `local` altogether, your `local` can be a global composer installation or phar file somewhere in PATH.

Comment (DanielRuf) [reply src/Codeception/Command/Run.php]: > This description could be better, what is that `local`?

A local version to a Composer-installed version.

Comment (DanielRuf) [reply src/Codeception/Command/Run.php]: `prefer-local` was suggested by @DavertMik

Comment (DavertMik) [reply src/Codeception/Command/Run.php]: @Naktibalda how would you call it?

File: autoload.php [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <?php
 
 $autoloadFile = './vendor/codeception/codeception/autoload.php';
-if (file_exists('./vendor/autoload.php') && file_exists($autoloadFile) && __FILE__ != realpath($autoloadFile)) {
+if (!isset($argv['--prefer-local']) && file_exists('./vendor/autoload.php') && file_exists($autoloadFile) && __FILE__ != realpath($autoloadFile)) {
```
Comment (Naktibalda) [inline autoload.php]: Does this condition even work? `$argv` isn't an associative array, it uses 0-based indexes.

File: src/Codeception/Command/Run.php (hunk 2 above)
Comment (Naktibalda) [reply src/Codeception/Command/Run.php]: From my point of view, codeception in `vendor` is as local as possible, the orther thing is less local.
`--no-redirect` is technical, but it is straitght to the point.
A suggestion to use `--no-redirect` option could be added to `Redirecting to Composer-installed version in vendor/codeception` message.

File: autoload.php (hunk 3 above)
Comment (Naktibalda) [reply autoload.php]: Please test this change in your own project, install Codeception from this branch and check if it works as you expect it to.
There is no point to merge it if it turns out that it doesn't solve your issue.

Comment (DanielRuf) [reply autoload.php]: I have changed this.

File: src/Codeception/Command/Run.php (hunk 2 above)
Comment (DavertMik) [reply src/Codeception/Command/Run.php]: Redirect can be confused with browser redirection. 
So I'd like to have something better here. 

`--prefer-current`
`--no-vendor`

**Comment by DavertMik** [comment]
Currently tests fail with 

```
--> Notice: Undefined variable: argv in /home/runner/Codeception/autoload.php on line 4
```

**Comment by DanielRuf** [comment]
> Currently tests fail with
> 
> ```
> --> Notice: Undefined variable: argv in /home/runner/Codeception/autoload.php on line 4
> ```

Not failing due to this but produces some tests to create a different output.
I have added a better check.

**Comment by DavertMik** [comment]
@Naktibalda are we ready to merge this?

**Comment by Naktibalda** [comment]
I have tried to use this feature and got the error I expected on a second run:

```
/c/Projects/codeception-zend-expressive-tests$ php ../Codeception/codecept run functional --no-redirect :ClickCept -vvv
Codeception PHP Testing Framework v3.1.0
Powered by PHPUnit 7.5.1 by Sebastian Bergmann and contributors.
Running with seed:


  init app in _initialize
  start initApp
  end initApp
  initialized response collector

Functional Tests (1) ------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Modules: ZendExpressive, \Helper\Functional, REST
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ClickCept: Click a link and see a change in url
Signature: ClickCept
Test: tests/functional/ClickCept.php
Scenario --
  reuse same app in _before
  initialized response collector
 I am on page "/"
  [Request Headers] []
  use old app app
  older version
 ERROR

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 2.34 seconds, Memory: 12.00MB

There was 1 error:

---------
1) ClickCept: Click a link and see a change in url
 Test  tests/functional/ClickCept.php

  [Error] Call to undefined method Symfony\Component\BrowserKit\Response::getStatusCode()


Scenario Steps:

 1. $I->amOnPage("/") at tests/functional/ClickCept.php:5

/c/Projects/Codeception/vendor/symfony/browser-kit/Client.php:414
/c/Projects/Codeception/src/Codeception/Lib/InnerBrowser.php:209
/c/Projects/Codeception/src/Codeception/Lib/InnerBrowser.php:257
/c/Projects/Codeception/src/Codeception/Lib/InnerBrowser.php:361
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/codeception/src/Codeception/Step.php:265
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/codeception/src/Codeception/Scenario.php:72
/c/Projects/codeception-zend-expressive-tests/tests/_support/_generated/FunctionalTesterActions.php:99
/c/Projects/codeception-zend-expressive-tests/tests/functional/ClickCept.php:5
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/codeception/src/Codeception/Test/Cept.php:42
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/codeception/src/Codeception/Test/Test.php:89
/c/Projects/codeception-zend-expressive-tests/vendor/phpunit/phpunit/src/Framework/TestSuite.php:746
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/phpunit-wrapper/src/Runner.php:117
/c/Projects/codeception-zend-expressive-tests/vendor/codeception/codeception/src/Codeception/SuiteManager.php:158
/c/Projects/Codeception/src/Codeception/Codecept.php:193
/c/Projects/Codeception/src/Codeception/Codecept.php:160
/c/Projects/Codeception/src/Codeception/Command/Run.php:502
/c/Projects/Codeception/src/Codeception/Command/Run.php:397
/c/Projects/Codeception/vendor/symfony/console/Command/Command.php:255
/c/Projects/Codeception/vendor/symfony/console/Application.php:921
/c/Projects/Codeception/vendor/symfony/console/Application.php:273
/c/Projects/Codeception/vendor/symfony/console/Application.php:149
/c/Projects/Codeception/src/Codeception/Application.php:107
/c/Projects/Codeception/codecept:43


ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

Please notice that some Codeception classes are loaded from one vendor directory, and some from other.

Disclaimer: use it at your own risk :)

**Comment by DanielRuf** [comment]
> Disclaimer: use it at your own risk :)

Definitely, this is why it is optional and oonly used when needed and not the default case.

**Pull request merged**